File: src/harness/registry.py
"""Adapter registry — dispatches RunSpec.harness to the right Adapter.

Shipped adapters are registered when the `harness` package is imported
(`harness/__init__.py` imports `harness.adapters`), so the registry is
populated immediately after `import harness`.
"""
from __future__ import annotations

from harness._subproc import SubprocOutcome
from harness.base import (
    Adapter,
    Backend,
    BuildCommand,
    Capabilities,
    HarnessError,
    ParsedOutput,
    RunResult,
    RunSpec,
    validate_backend,
)

_REGISTRY: dict[str, type[Adapter]] = {}


def register(name: str, adapter_cls: type[Adapter]) -> None:
    """Register `adapter_cls` under `name`. Re-registering the same class is a
    no-op; a different class under an existing name raises `duplicate-adapter`.
    """
    existing = _REGISTRY.get(name)
    if existing is not None and existing is not adapter_cls:
        raise HarnessError(f"adapter name collision: {name!r} is already registered", code="duplicate-adapter")
    _REGISTRY[name] = adapter_cls


def list_adapters() -> list[str]:
    """Registered harness names, sorted by code point (locale-independent)."""
    return sorted(_REGISTRY)


def _adapter_class(name: str) -> type[Adapter]:
    try:
        return _REGISTRY[name]
    except KeyError:
        available = ", ".join(list_adapters()) or "(none registered)"
        raise HarnessError(f"unknown harness {name!r}; available: {available}", code="unknown-harness") from None


def get_adapter(name: str) -> Adapter:
    return _adapter_class(name)()


def get_capabilities(name: str, backend: Backend = "cli") -> Capabilities:
    """What `name` supports on `backend` in this release. Pure; no CLI probe.

    Raises `unknown-harness` for unregistered names and `unsupported-backend`
    / `invalid-options` for backends that cannot be queried honestly.
    """
    adapter_cls = _adapter_class(name)
    validate_backend(backend)
    policies = ("upstream", "bypass") if adapter_cls.permission_bypass_args is not None else ("upstream",)
    return Capabilities(
        backend="cli",
        permission_policies=policies,
        native_options=adapter_cls.native_options_kind,
        streaming=True,
        cancellation=True,
        sessions=False,
        config_home_env=adapter_cls.config_home_env,
        config_file_flag=adapter_cls.config_file_flag,
    )


def _validated_adapter(spec: RunSpec) -> Adapter:
    """Resolve `spec.harness` and reject unsupported backend/policy/options.

    Runs even for custom adapters whose `build_command` skips validation, so
    every registry entrypoint refuses rpc/sdk and invalid options up front.
    """
    adapter = get_adapter(spec.harness)
    adapter.validate_run_spec(spec)
    return adapter


def build_command(spec: RunSpec) -> BuildCommand:
    """Build the subprocess command without executing it. Pure: nothing is
    written; `prepare_command` applies the plan."""
    adapter = _validated_adapter(spec)
    return adapter._finalized(spec, adapter.build_command(spec))


def parse_output(spec: RunSpec, outcome: SubprocOutcome) -> ParsedOutput:
    """Parse adapter output after execution."""
    return _validated_adapter(spec).parse_output(spec, outcome)


def run(spec: RunSpec) -> RunResult:
    """Full headless invocation: build_command + exec + parse_output."""
    return _validated_adapter(spec).run(spec)


async def run_async(spec: RunSpec) -> RunResult:
    """Non-blocking headless invocation: build_command + async exec + parse_output."""
    return await _validated_adapter(spec).run_async(spec)

File: ts/src/registry.ts
import type { Adapter, Backend, BuildCommand, Capabilities, ParsedOutput, RunResult, RunSpec, SubprocOutcome } from './base.js'
import { HarnessError, finalizeCommand, resolveBackend, validateRunSpec } from './base.js'
import { cleanupCommand, prepareCommand } from './instructions.js'
import { describeError, prepareLaunch, runLifecycle } from './lifecycle.js'

const registry = new Map<string, Adapter>()

/** Re-registering the same adapter object under the same name is a no-op; any other collision fails. */
export function register(name: string, adapter: Adapter): void {
  const existing = registry.get(name)
  if (existing === adapter) return
  if (existing) {
    throw new HarnessError(`Duplicate adapter registration: "${name}"`, 'duplicate-adapter')
  }
  registry.set(name, adapter)
}

export function listAdapters(): string[] {
  // Default sort is locale-independent UTF-16 code-unit order.
  return Array.from(registry.keys()).sort()
}

export function getAdapter(name: string): Adapter {
  const adapter = registry.get(name)
  if (!adapter) {
    throw new HarnessError(`Unknown harness: "${name}". Available: ${listAdapters().join(', ')}`, 'unknown-harness')
  }
  return adapter
}

/**
 * Static support report for a harness on a backend. Reports what the adapter
 * implements, not whether the CLI is installed or authenticated. Unsupported
 * backends reject rather than returning a fabricated capability set.
 */
export function getCapabilities(name: string, backend: Backend = 'cli'): Capabilities {
  const adapter = getAdapter(name)
  return {
    backend: resolveBackend(backend),
    permissionPolicies: adapter.permissionBypassArgs ? ['upstream', 'bypass'] : ['upstream'],
    nativeOptions: adapter.nativeOptionsKind ?? null,
    configHomeEnv: adapter.configHomeEnv ?? null,
    configFileFlag: adapter.configFileFlag ?? null,
    streaming: true,
    cancellation: true,
    sessions: false,
  }
}

/**
 * Validate, dispatch, and pass the adapter's command back through the shared
 * finalizer. Idempotent for built-in adapters; it gives minimal third-party
 * adapters the same executable/config-home/env layering and absolute cwd.
 */
export function buildCommand(spec: RunSpec): BuildCommand {
  const adapter = getAdapter(spec.harness)
  const validated = validateRunSpec(adapter, spec)
  return finalizeCommand(adapter, spec, validated, adapter.buildCommand(spec))
}

export function parseOutput(spec: RunSpec, outcome: SubprocOutcome): ParsedOutput {
  const adapter = getAdapter(spec.harness)
  validateRunSpec(adapter, spec)
  return adapter.parseOutput(spec, outcome)
}

/**
 * Caller-owned spec and env are copied before command planning and execution.
 * Run I/O options are validated before the workdir is touched, so a bad
 * option never leaves a lease behind. A parser exception becomes
 * `parseError` with null metrics; the process outcome is kept.
 */
async function execute(spec: RunSpec): Promise<RunResult> {
  const frozen = spec.env === undefined ? { ...spec } : { ...spec, env: { ...spec.env } }
  const adapter = getAdapter(frozen.harness)
  const built = buildCommand(frozen)
  const request = prepareLaunch([built.cmd, ...built.args], {
    cwd: built.cwd,
    timeoutSeconds: frozen.timeoutSeconds,
    inactivityTimeoutSeconds: frozen.inactivityTimeoutSeconds,
    maxOutputBytes: frozen.maxOutputBytes,
    stdin: frozen.stdin,
    onOutput: frozen.onOutput,
    extraEnv: built.env,
    cancel: frozen.cancel,
    gracefulSignal: built.gracefulSignal,
  })
  const prepared = prepareCommand(built)
  let cleanupSafe = false
  try {
    // Only a resolved outcome proves the group is stopped; a rejection here is a teardown failure.
    const outcome = await runLifecycle(request, frozen.cancel, undefined, frozen.onOutput)
    cleanupSafe = true
    let parsed: ParsedOutput = { costUsd: null, tokensIn: null, tokensOut: null, raw: null }
    let parseError: string | null = null
    try {
      parsed = adapter.parseOutput(frozen, outcome)
    } catch (error) {
      parseError = describeError(error)
    }
    return {
      harness: frozen.harness,
      model: built.model === undefined ? frozen.model || adapter.defaultModel : built.model,
      exitCode: outcome.exitCode,
      durationSeconds: outcome.durationSeconds,
      stdout: outcome.stdout,
      stderr: outcome.stderr,
      timedOut: outcome.timedOut,
      termination: outcome.termination,
      signal: outcome.signal,
      launchError: outcome.launchError,
      stdoutBytes: outcome.stdoutBytes,
      stderrBytes: outcome.stderrBytes,
      stdoutTruncated: outcome.stdoutTruncated,
      stderrTruncated: outcome.stderrTruncated,
      callbackError: outcome.callbackError,
      timeoutKind: outcome.timeoutKind,
      parseError,
      costUsd: parsed.costUsd,
      tokensIn: parsed.tokensIn,
      tokensOut: parsed.tokensOut,
      raw: parsed.raw,
    }
  } finally {
    if (cleanupSafe) cleanupCommand(prepared)
  }
}

/**
 * Full headless invocation. Both entry points execute in-process on the
 * async engine (no event-loop blocking), so `spec.cancel` can abort either.
 */
export function run(spec: RunSpec): Promise<RunResult> {
  return execute(spec)
}

export function runAsync(spec: RunSpec): Promise<RunResult> {
  return execute(spec)
}
